docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,548 @@
+- Does not currently support foreign keys for tables with user-defined partitioning. This includes both reference and child tables.
+- A foreign key constraint cannot reference a virtual generated column, but stored generated column is ok.
+
+#### Handle In DDL Owner
+
+When DDL Owner handle create table job, DDL owner need to create a new table, and update the reference tables information.
+
+At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create new table and 
+update all reference tables in one schema version, since this may break foreign key constrain, such as delete reference table
+without foreign key constrain check in child table.
+
+```sql
+-- In TiDB-1 and Schema Version is 1
+insert into t_has_foreign_key values (1, 1);
+
+-- In TiDB-0 and  Schema Version is 0
+delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
+```
+
+So, when create a table with foreign key, we need multi-schema version change:
+
+1. None -> Write Only: Create table with state is `write-only`, and update all reference tables info.
+2. Write Only -> Done: Update the new created table state to `public`.
+
+In step-1, we need update some table info in one schema-version. Technically, we can implement is since we already support `ActionCreateTables` DDL job.
+
+### Alter Table Add Foreign Key
+
+Here is an example:
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int);
+alter  table t2 add foreign key fk(a) references t1(id) ON DELETE CASCADE;
+```
+
+Just like create table, we should validate first, and return error if the conditions for creating foreign keys are not met, and also need to do double-check.
+
+When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
+And this is divides the problem into two cases:
+- Case-1: No need to auto create index, and only add foreign key constrain.
+- Case-2: Need auto create index for foreign key
+
+#### Case-1: Only add foreign key constrain
+
+The DDL owner handle add foreign key constrain step is:
+
+1. None -> Write Only: add foreign key constrain which state is `write-only` into table and update the reference table info.
+3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
+   ```sql
+   select count(*) from t2 where t2.a not in (select id from t1);
+   ```
+   The expected result is `0`, otherwise, an error is returned and cancel the ddl job.
+4. Write Reorg -> Public: update the foreign key constrain state to `public`.
+
+DML should also check the foreign key constrain which state is `write-only`
+
+#### Case-2: Auto create index for foreign key and add foreign key constrain
+
+As TiDB support multi-schema change now, we can split this into 2 sub-ddl job.
+- Add Index DDL job
```

Comment:
The index added here is public to use?

Reply:
Nice catch. Only when the added index is ready to use(finish add index reorg, but the state is not public), then we can start to add the foreign key constrain.